File: 프로그래머스/2/134239. 우박수열 정적분/우박수열 정적분.js
function collatz(k) {
    const arr = [k];
    while (k !== 1) {
        k = k % 2 === 0 ? k / 2 : k * 3 + 1;
        arr.push(k);
    }
    
    return arr;
}

function solution(k, ranges) {
    let answer = [];
    const ubak = collatz(k);
    let area = [];

    for (let i = 0; i < ubak.length - 1; i++) area.push((ubak[i] + ubak[i + 1]) / 2);

    const cumulativeArea = [0];
    for (let i = 0; i < area.length; i++) {
        cumulativeArea.push(cumulativeArea[i] + area[i]);
    }

    const n = cumulativeArea.length - 1;

    for (const [a, b] of ranges) {
        const endIndex = n + b;
        if (a > endIndex) answer.push(-1);
        else if (a === endIndex) answer.push(0);
        else answer.push(cumulativeArea[endIndex] - cumulativeArea[a]);
    }

    return answer;
}

File: 프로그래머스/2/134239. 우박수열 정적분/우박수열 정적분.py
def getCollatz(k):
    result = []
    
    while k != 1:
        result.append(k)
        k = k / 2 if k % 2 == 0 else k * 3 + 1     
    result.append(k)
    
    return result
 
def solution(k, ranges):
    answer = []
    collatz = getCollatz(k)

    for r in ranges:
        total = 0
        ubakRange = collatz[r[0] : len(collatz)+r[1]]
        
        if r[0] >= r[1] + len(collatz):
            answer.append(-1)
            continue
            
        for i in range(len(ubakRange) - 1):
            total += (((ubakRange[i] + ubakRange[i+1]) * 1) / 2)
            
        answer.append(total)
        
    return answer